File: api/db/supabase.py
from typing import Any
from supabase import create_client, Client
from .interface import DatabaseClient


class SupabaseClient(DatabaseClient):
    def __init__(self, url: str, key: str) -> None:
        self._client: Client = create_client(url, key)

    def _apply_filters(self, query, filters: dict[str, Any]):
        for column, value in filters.items():
            query = query.eq(column, value)
        return query

    async def select(
        self,
        table: str,
        *,
        columns: str = "*",
        filters: dict[str, Any] | None = None,
        limit: int | None = None,
    ) -> list[dict]:
        query = self._client.table(table).select(columns)
        if filters:
            query = self._apply_filters(query, filters)
        if limit:
            query = query.limit(limit)
        return query.execute().data

    async def insert(
        self,
        table: str,
        data: dict | list[dict],
    ) -> list[dict]:
        return self._client.table(table).insert(data).execute().data

    async def update(
        self,
        table: str,
        data: dict,
        *,
        filters: dict[str, Any],
    ) -> list[dict]:
        query = self._client.table(table).update(data)
        query = self._apply_filters(query, filters)
        return query.execute().data

    async def delete(
        self,
        table: str,
        *,
        filters: dict[str, Any],
    ) -> list[dict]:
        query = self._client.table(table).delete()
        query = self._apply_filters(query, filters)
        return query.execute().data

    async def rpc(
        self,
        function_name: str,
        params: dict | None = None,
    ) -> Any:
        return self._client.rpc(function_name, params or {}).execute().data

File: src/lib/db/supabase.js
import { createClient } from '@supabase/supabase-js'
import { DatabaseClient } from './interface'

export class SupabaseClient extends DatabaseClient {
  constructor() {
    super()
    this._client = createClient(
      import.meta.env.VITE_SUPABASE_URL,
      import.meta.env.VITE_SUPABASE_ANON_KEY,
    )
  }

  _applyFilters(query, filters = {}) {
    return Object.entries(filters).reduce((q, [col, val]) => q.eq(col, val), query)
  }

  async select(table, { columns = '*', filters, limit } = {}) {
    let q = this._client.from(table).select(columns)
    if (filters) q = this._applyFilters(q, filters)
    if (limit)   q = q.limit(limit)
    const { data, error } = await q
    if (error) throw error
    return data
  }

  async insert(table, data) {
    const { data: rows, error } = await this._client.from(table).insert(data).select()
    if (error) throw error
    return rows
  }

  async update(table, data, filters) {
    let q = this._client.from(table).update(data)
    q = this._applyFilters(q, filters)
    const { data: rows, error } = await q.select()
    if (error) throw error
    return rows
  }

  async delete(table, filters) {
    let q = this._client.from(table).delete()
    q = this._applyFilters(q, filters)
    const { data: rows, error } = await q.select()
    if (error) throw error
    return rows
  }

  async rpc(fn, params = {}) {
    const { data, error } = await this._client.rpc(fn, params)
    if (error) throw error
    return data
  }

  /** Expose raw client for auth, realtime, storage, etc. */
  get raw() { return this._client }
}
